### ISSUE TITLE
Editing websocket messages throws exception and kills connection

### ISSUE THREAD

**Issue by juhakivekas** [zaproxy/zaproxy#4149]
When editing binary websockets with an extender script the underlying proxy fails:
```
3886033 [ZAP-WS-Listener (local) 'xxx.xxx.com:443 (#6)'] ERROR org.zaproxy.zap.ZAP$UncaughtExceptionLogger  - Exception in thread "ZAP-WS-Listener (local) 'xxx.xxx.com:443 (#6)'"
java.nio.BufferUnderflowException
	at java.nio.HeapByteBuffer.get(HeapByteBuffer.java:151)
	at org.zaproxy.zap.extension.websocket.WebSocketProxyV13$WebSocketMessageV13.forward(WebSocketProxyV13.java:618)
	at org.zaproxy.zap.extension.websocket.WebSocketProxy.processRead(WebSocketProxy.java:564)
	at org.zaproxy.zap.extension.websocket.WebSocketListener.run(WebSocketListener.java:89)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```
The same behaviour is repriducible in the manual hex message editor. This also makes a blank ZAP error message box appear.
```
5127571 [Thread-439] WARN org.parosproxy.paros.extension.manualrequest.ManualRequestEditorDialog  - 
java.nio.BufferUnderflowException
	at java.nio.HeapByteBuffer.get(HeapByteBuffer.java:151)
	at org.zaproxy.zap.extension.websocket.WebSocketProxyV13$WebSocketMessageV13.forward(WebSocketProxyV13.java:618)
	at org.zaproxy.zap.extension.websocket.WebSocketProxy.sendAndNotify(WebSocketProxy.java:898)
	at org.zaproxy.zap.extension.websocket.manualsend.WebSocketPanelSender.handleSendMessage(WebSocketPanelSender.java:74)
	at org.parosproxy.paros.extension.manualrequest.ManualRequestEditorDialog$3.run(ManualRequestEditorDialog.java:211)
	at java.lang.Thread.run(Thread.java:748)

```
Original causing script:
```
var Hex = Java.type("org.apache.commons.codec.binary.Hex")

var tamperListener = new org.zaproxy.zap.extension.websocket.WebSocketSenderListener(){
	getListenerOrder: function(){
		return 5;
	},
	onMessageFrame: function(channelId, message, initiator){
		if(message.getOpcodeString() != "BINARY" || message.getDirection() != org.zaproxy.zap.extension.websocket.WebSocketMessage.Direction.OUTGOING){
			return
		}
		
		html = 'FOOBAR'
		injection  = Hex.encodeHexString(html.getBytes())
		marker     = Hex.encodeHexString('INJECT'.getBytes())
		hexpayload = Hex.encodeHexString(message.getPayload())
		if(hexpayload.contains(marker)){
			hexpayload = hexpayload.replace(marker, injection)
			message.setPayload(Hex.decodeHex(hexpayload.toCharArray()))
		}
	},
	onStateChange: function(state, proxy){
	}
}

/**
 * This function is called when the script is enabled, it adds the custom WebSocketSenderListener object to the WebSocket extension.
 */
function install(helper) {
	extensionWebSocket = org.parosproxy.paros.control.Control.getSingleton().getExtensionLoader().getExtension(
		org.zaproxy.zap.extension.websocket.ExtensionWebSocket.NAME)
	extensionWebSocket.addAllChannelSenderListener(tamperListener)
	//note that the listener will not be attached to already existing WebSocket connections
	//reloading pages to reconnect all sockets might be needed to enable the added listener
}

/**
 * This function is called when the script is disabled, it removes the custom WebSocketSenderListener object to the WebSocket extension.
 */
function uninstall(helper) {
	extensionWebSocket = org.parosproxy.paros.control.Control.getSingleton().getExtensionLoader().getExtension(
		org.zaproxy.zap.extension.websocket.ExtensionWebSocket.NAME)
	extensionWebSocket.removeAllChannelSenderListener(tamperListener)
}
```

ZAP 2.7.0
WebSockets 14
Debian 9

**Comment by thc202** [comment]
This happens with all types of messages (binary, text), after obtain, change, and send the message. The following patch addresses the issue:
```diff
diff --git a/src/org/zaproxy/zap/extension/websocket/WebSocketProxyV13.java b/src/org/zaproxy/zap/extension/websocket/WebSocketProxyV13.java
index f72104060..902767d3c 100644
--- a/src/org/zaproxy/zap/extension/websocket/WebSocketProxyV13.java
+++ b/src/org/zaproxy/zap/extension/websocket/WebSocketProxyV13.java
@@ -671,6 +671,7 @@ public class WebSocketProxyV13 extends WebSocketProxy {
                        payload.rewind();
                        byte[] bytes = new byte[payload.limit()];
                        payload.get(bytes);
+                       payload.rewind();
                        return bytes;
                }
```

**Comment by thc202** [comment]
Fixed in zaproxy/zap-extensions#1362.

**Comment by thc202** [comment]
Released in version 15 of WebSockets add-on.
